packages/react-reconciler/src/ReactFiber.js

```diff
@@ -431,6 +436,25 @@ export function createWorkInProgress(
     workInProgress.treeBaseDuration = current.treeBaseDuration;
   }
 
+  if (__DEV__) {
+    switch (workInProgress.tag) {
+      case IndeterminateComponent:
+      case FunctionComponent:
+      case SimpleMemoComponent:
+        workInProgress.type = resolveFunctionForHotReloading(
+          workInProgress.type,
```

Comment:
I believe this should be `current.type`.

I think this answers my question around simple memo. You're not getting a new `type` from `element.type` but reading it from the fiber which is the underlying render function. I would've expected you to read the type again from element but I guess this is fine too. It just means that if we ever start getting updatable types during reconciliation this would break but we can deal with that then.